internal/app/singularity/verify_ocsp.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2022, Sylabs Inc. All rights reserved.
+// Copyright (c) 2020-202, ICS-FORTH.  All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the LICENSE.md file
+// distributed with the sources of this project regarding your rights to use or distribute this
+// software.
+
+package singularity
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/sylabs/singularity/pkg/sylog"
+	"golang.org/x/crypto/ocsp"
+)
+
+/*
+Online Certificate Status Protocol - OCSP
+
+OCSP responder is used to provide real-time verification of the revocation status of an X.509 certificate.
+RFC: https://www.rfc-editor.org/rfc/rfc6960
+*/
+
+const (
+	PKIXOCSPNoCheck = "1.3.6.1.5.5.7.48.1.5"
+)
+
+var errOCSP = errors.New("OCSP verification has failed")
+
+func OCSPVerify(chain ...*x509.Certificate) error {
+	// use the pool as an index for certificate issuers.
+	// fixme: we can drop this lookup if we assume that certificate N is always signed by certificate N+1.
+	pool := map[string]*x509.Certificate{}
+
+	for _, cert := range chain {
+		pool[string(cert.SubjectKeyId)] = cert
+	}
+
+	// recursively validate the certificate chain
+	for _, cert := range chain {
+		if err := validateCertificate(cert, pool); err != nil {
+			sylog.Warningf("OCSP verification has failed. Err: %s", err)
+			return errOCSP
+		}
+	}
+
+	return nil
+}
+
+func validateCertificate(cert *x509.Certificate, pool map[string]*x509.Certificate) error {
+	if len(cert.AuthorityKeyId) == 0 || string(cert.SubjectKeyId) == string(cert.AuthorityKeyId) {
+		sylog.Infof("skip self-signed certificate (%s)", cert.Subject.String())
+
+		return nil
+	}
+
+	/*---------------------------------------------------
+	 * Retrieve the CA who issued the certificate in question.
+	 *---------------------------------------------------*/
+
+	// firstly, look for the issuer in the pool of certificates.
+	issuer, exists := pool[string(cert.AuthorityKeyId)]
+	if !exists {
+		// if not found anywhere locally, try to download it
+		missingCerts, err := downloadCertsFromURLs(cert.IssuingCertificateURL...)
+		if err != nil {
+			return errors.Wrapf(err, "download cert error")
```

Comment:
Applicable in several places... use wrapped errors from the standard library, ex:

```suggestion
			return fmt.Errorf("download cert error: %w", err)
```